File: 0624/solution.go
package main

func maxDistance(arrays [][]int) int {
	res := 0
	prevBig := arrays[0][len(arrays[0])-1]
	prevSmall := arrays[0][0]

	for i := 1; i < len(arrays); i++ {
		currBig := arrays[i][len(arrays[i])-1]
		currSmall := arrays[i][0]
		res = max(res, currBig-prevSmall)
		res = max(res, prevBig-currSmall)
		prevBig = max(prevBig, currBig)
		prevSmall = min(prevSmall, currSmall)
	}
	return res
}

func min(a int, b int) int {
	if a < b {
		return a
	} else {
		return b
	}
}

func max(a int, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

File: 0624/solution.java
import java.util.List;

class Solution {
    public int maxDistance(List<List<Integer>> arrays) {
        int res = 0;
        int prevBig = arrays.get(0).get(arrays.get(0).size() - 1);
        int prevSmall = arrays.get(0).get(0);

        for (int i = 1; i < arrays.size(); i++) {
            int currBig = arrays.get(i).get(arrays.get(i).size() - 1);
            int currSmall = arrays.get(i).get(0);
            res = Math.max(res, currBig - prevSmall);
            res = Math.max(res, prevBig - currSmall);
            prevBig = Math.max(prevBig, currBig);
            prevSmall = Math.min(prevSmall, currSmall);
        }
        return res;
    }
}

// time O(n)
// space O(1)